src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -41,26 +40,61 @@ public Product getProduct(Long id) {
             return products.stream()
                     .filter(product -> product.getId().equals(id))
                     .findFirst()
-                    .orElseThrow(() -> new ProductNotFoundException("Product not found"));
+                    .orElseThrow(() -> new ProductNotFoundException(id));
         } finally {
             reentrantReadWriteLock.readLock().unlock();
         }
     }
 
     @Override
-    public List<Product> findProducts() {
+    public List<Product> findProducts(final String query, String sortField, String sortOrder) {
         reentrantReadWriteLock.readLock().lock();
-
         try {
-            return products.stream().
-                    filter(product -> product.getStock() > 0).
-                    filter(product -> product.getPrice() != null).
-                    collect(Collectors.toList());
+            List<String> keywords = (query == null || query.isBlank())
+                    ? List.of()
+                    : Arrays.asList(query.toLowerCase().split("\\s+"));
+
+            Comparator<Product> relevanceComparator = this.getRelevanceComparator(keywords, query);
+            Comparator<Product> productComparator = this.getProductComparator(sortField, sortOrder);
+
+            return products.stream()
+                    .filter(product -> product.getStock() > 0 && product.getPrice() != null)
+                    .filter(product -> keywords.isEmpty() || keywords.stream()
+                            .anyMatch(word -> product.getDescription().toLowerCase().contains(word)))
+                    .sorted(relevanceComparator.thenComparing(productComparator))
+                    .collect(Collectors.toList());
         } finally {
             reentrantReadWriteLock.readLock().unlock();
         }
     }
 
+    private Comparator<Product> getProductComparator(String sortField, String sortOrder) {
+        if (sortField == null) {
+            return (p1, p2) -> 0;
+        }
+
+        Comparator<Product> comparator = switch (sortField.toLowerCase()) {
+            case "price" -> Comparator.comparing(Product::getPrice, Comparator.nullsLast(Comparator.naturalOrder()));
+            case "description" ->
+                    Comparator.comparing(Product::getDescription, Comparator.nullsLast(Comparator.naturalOrder()));
+            default -> (p1, p2) -> 0;
+        };
+
+        return "desc".equalsIgnoreCase(sortOrder) ? comparator.reversed() : comparator;
+    }
+
+    private Comparator<Product> getRelevanceComparator(List<String> keywords, String query) {
+        if (query != null) {
+            return (p1, p2) -> 0;
+        }
+
+        return Comparator.comparingInt((Product product) ->
+                (int) keywords.stream()
+                        .filter(word -> product.getDescription().toLowerCase().contains(word))
+                        .count()
+        ).reversed();
```

Comment:
here we're comparing also based on number of words in the description, it means - if we search "samsung galaxy s", then we should get "samsung galaxy s" and then "samsung galaxy s 3"